shared/src/balancer_event_handler.rs

```diff
@@ -3,195 +3,183 @@ use crate::{
 };
 use tokio::sync::Mutex;
 
-#[derive(Clone, Debug)]
-pub enum BalancerEvent {
-    PoolRegistered(PoolRegistered),
-    TokensRegistered(TokensRegistered),
-}
-
-#[derive(Debug, Clone, PartialEq)]
-pub struct PoolRegistered {
-    pub pool_id: H256,
+#[derive(Copy, Debug, Default, Clone, Eq, PartialEq, Hash)]
+pub struct PoolCreated {
     pub pool_address: H160,
-    pub specialization: PoolSpecialization,
-}
-
-#[derive(Debug, Clone, PartialEq)]
-pub struct TokensRegistered {
-    pub pool_id: H256,
-    pub tokens: Vec<H160>,
 }
 
 #[derive(Clone, Debug, Eq, PartialEq, Hash)]
 pub struct WeightedPool {
     pool_id: H256,
     pool_address: H160,
     tokens: Vec<H160>,
-    specialization: PoolSpecialization,
+    normalized_weights: Vec<U256>,
     block_created: u64,
 }
 
-#[derive(Clone, Debug, Default, PartialEq)]
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Hash)]
 pub struct WeightedPoolBuilder {
-    pool_registration: Option<PoolRegistered>,
-    tokens_registration: Option<TokensRegistered>,
-    /// Both Pool and Tokens Registered events are emitted in the same transaction
-    /// https://github.com/balancer-labs/balancer-v2-monorepo/blob/70843e6a61ad11208c1cfabf5cfe15be216ca8d3/pkg/pool-utils/contracts/BasePool.sol#L128-L130
+    pool_creation: PoolCreated,
     /// block_number is only contained in the EventIndex
     block_created: u64,
 }
 
 impl WeightedPoolBuilder {
-    fn into_pool(self) -> Option<WeightedPool> {
-        if let (Some(pool_registration), Some(tokens_registration)) =
-            (self.pool_registration, self.tokens_registration)
-        {
-            return Some(WeightedPool {
-                pool_id: pool_registration.pool_id,
-                pool_address: pool_registration.pool_address,
-                tokens: tokens_registration.tokens,
-                specialization: pool_registration.specialization,
-                block_created: self.block_created,
-            });
-        }
-        None
+    async fn into_pool(self, data_fetcher: &dyn PoolDataFetching) -> Result<WeightedPool> {
+        let pool_address = self.pool_creation.pool_address;
+        let pool_data = data_fetcher.get_pool_data(pool_address).await?;
+        return Ok(WeightedPool {
+            pool_id: pool_data.pool_id,
+            pool_address,
+            tokens: pool_data.tokens,
+            normalized_weights: pool_data.weights,
+            block_created: self.block_created,
+        });
+    }
+}
+
+#[derive(Clone)]
+pub struct WeightedPoolData {
+    pool_id: H256,
+    tokens: Vec<H160>,
+    weights: Vec<U256>,
+}
+
+#[automock]
+#[async_trait::async_trait]
+trait PoolDataFetching: Send + Sync {
+    async fn get_normalized_weights(&self, pool_address: H160) -> Result<Vec<U256>>;
+    async fn get_pool_tokens(&self, pool_id: H256) -> Result<Vec<H160>>;
+    async fn get_pool_id(&self, pool_address: H160) -> Result<H256>;
+    async fn get_pool_data(&self, pool_address: H160) -> Result<WeightedPoolData>;
```

Comment:
```suggestion
    async fn get_pool_data(&self, pool_address: H160) -> Result<WeightedPoolData>;
```

Since we intend to get the data "all or nothing", I think it makes more sense for this to be a single method. It would also allow you to make a batch call to read all the pool data together (instead of individual calls).